1-producer-consumer/main.go

```diff
@@ -11,38 +11,48 @@ package main
 import (
 	"fmt"
 	"time"
+  "sync"
 )
 
-func producer(stream Stream) (tweets []*Tweet) {
-	for {
+func producer(stream Stream, tweets chan<- Tweet, wg *sync.WaitGroup) {
+  defer wg.Done()	
+  for {
 		tweet, err := stream.Next()
 		if err == ErrEOF {
-			return tweets
+      tweets <- Tweet{"","done"}
+      wg.Done()	
 		}
-
-		tweets = append(tweets, tweet)
+		tweets <- *tweet
 	}
 }
 
-func consumer(tweets []*Tweet) {
-	for _, t := range tweets {
+func consumer(tweets <-chan Tweet, wg *sync.WaitGroup) {
+  for t := range tweets {
+    if (t.Username == "" && t.Text == "") {  
+      wg.Done()
+      return 
+    } 
 		if t.IsTalkingAboutGo() {
-			fmt.Println(t.Username, "\ttweets about golang")
-		} else {
+      fmt.Println(t.Username, "\ttweets about golang")
+		} else { 
 			fmt.Println(t.Username, "\tdoes not tweet about golang")
-		}
-	}
+    }	
+  }
 }
 
 func main() {
 	start := time.Now()
 	stream := GetMockStream()
-
+  tweets := make(chan Tweet)
+  var wg sync.WaitGroup
+  wg.Add(2) 
+  
 	// Producer
-	tweets := producer(stream)
+	go producer(stream, tweets, &wg)
 
 	// Consumer
-	consumer(tweets)
+	go consumer(tweets, &wg)
```

Comment:
Just a comment. You might not need to run the consumer concurrently. And maybe this allows you to get rid of the `WaitGroup`!